codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java

```diff
@@ -529,11 +527,11 @@ private boolean shouldIgnoreHeadersOrDataFrame(ChannelHandlerContext ctx, int st
                             ctx.channel(), frameName, streamId);
                     return true;
                 }
-                // Its possible that this frame would result in stream ID out of order creation (PROTOCOL ERROR) and its
-                // also possible that this frame is received on a CLOSED stream (STREAM_CLOSED after a RST_STREAM is
-                // sent). We don't have enough information to know for sure, so we choose the lesser of the two errors.
-                throw streamError(streamId, STREAM_CLOSED, "Received %s frame for an unknown stream %d",
-                                  frameName, streamId);
+                // If the stream could have existed in the past we assume this is a frame sent by the remote
```

Comment:
Just to be sure we stay in sync with the remote peer we can always throw a `streamError` to inform the remote peer that we don't understand the current frame. This will help provide visibly in the event the remote/local get out of sync for some reason. For example the spec recommends it for the PUSH_PROMISE case and in general allows multiple RST_STREAM frames in extraneous scenarios.

> An endpoint might receive a PUSH_PROMISE frame after it sends
      RST_STREAM.  PUSH_PROMISE causes a stream to become "reserved"
      even if the associated stream has been reset.  Therefore, a
      RST_STREAM is needed to close an unwanted promised stream.

> However, an endpoint MAY send additional RST_STREAM
   frames if it receives frames on a closed stream after more than a
   round-trip time.  This behavior is permitted to deal with misbehaving
   implementations.

Some considerations:
- there should be some limit to how many RST_STREAMs we are willing to send in this scenario before either going into graceful close or eventually giving up and hard closing. For example we don't want to indefinitely send RST_STREAMs. However this heuristic can be implemented in a followup PR and doesn't change the existing behavior.
- The spec says the following, but in this case we have discarded the local state and don't know if we have already sent a RST_STREAM at this time ... which the spec allows sending multiple RST_STREAMs (as indicated above).
> An endpoint MUST ignore frames that it
      receives on closed streams after it has sent a RST_STREAM frame.
      An endpoint MAY choose to limit the period over which it ignores
      frames and treat frames that arrive after this time as being in
      error.


```java
if (connection.streamMayHaveExisted(streamId)) {
  throw streamError(streamId, STREAM_CLOSED, "Received %s frame for an unknown stream %d", frameName, streamId);
} else {
  throw connectionError(PROTOCOL_ERROR, "Stream %d does not exist", streamId);
}
```

Reply:
@Scottmitch per your second consideration (if I'm understanding correctly), that would negate the original purpose of the PR. We want to ignore any data frames that come after the reset and not propagate failures to the stream.

Reply:
I looked at the original issue and here is what the RFC says about the scenario when we have sent a RST_STREAM frame and receive a DATA frame:

https://tools.ietf.org/html/rfc7540#section-5.1
> An endpoint MUST ignore frames that it
      receives on closed streams after it has sent a RST_STREAM frame.
      An endpoint MAY choose to limit the period over which it ignores
      frames and treat frames that arrive after this time as being in
      error.

https://tools.ietf.org/html/rfc7540#section-6.1
> If a DATA frame is received
   whose stream is not in "open" or "half-closed (local)" state, the
   recipient MUST respond with a stream error (Section 5.4.2) of type
   STREAM_CLOSED.

https://tools.ietf.org/html/rfc7540#section-5.4.2
> A RST_STREAM is the last frame that an endpoint can send on a stream.
   The peer that sends the RST_STREAM frame MUST be prepared to receive
   any frames that were sent or enqueued for sending by the remote peer.
   These frames can be ignored, except where they modify connection
   state (such as the state maintained for header compression
   (Section 4.3) or flow control).

> Normally, an endpoint SHOULD NOT send more than one RST_STREAM frame
   for any stream.  However, an endpoint MAY send additional RST_STREAM
   frames if it receives frames on a closed stream after more than a
   round-trip time.  This behavior is permitted to deal with misbehaving
   implementations.

> To avoid looping, an endpoint MUST NOT send a RST_STREAM in response
   to a RST_STREAM frame.

Here is my interpretation of the above somewhat contradictory statements:
1. If state transitions in a coordinated fashion between peers into the CLOSED state and a DATA frame is received in the CLOSED state, then you MUST send a RST_STREAM.
2. Sending a RST_STREAM is an uncoordinated means of forcing the stream into the closed state, and therefore you may receive frames that are already in flight. This frames can be ignored if you are reasonably confident that the remote peer has not yet received the RST_STREAM, but you are also allowed to send multiple RST_STREAMs.
3. Receiving a RST_STREAM frame while the stream is in the CLOSED state should have no protocol impacts.

Given the above criteria my preferred way to handle the scenario is the following:
- If the stream may have existed ... tracking/estimating RTT time is error prone and can be expensive for what is in large part a boundary condition. The RFC permits sending multiple RST_STREAM frames, doing so will have no negative implications from a protocol perspective*. A simpler approach would be to continue to send the RST_STREAM frames which provides positive acknowledgement for good peers that this frame was not accepted, and have some heuristic to "fail hard" (e.g. connection error) to catch the case of "misbehaving peers".
- if the stream was never in existence ... protocol error
- reduce the log level in `io.grpc.netty.NettyServerHandler` for this case (or consider a more selective log strategy)

(*) of course this will consume network resources but it is viewed as a tradeoff to keep things simple from state management and heuristic perspective. If this is proven to consume non-negligible network resources for some use cases I would first like to explore handling the heuristic of "not sending the RST_STREAM bcz I think this falls within the RTT of the previous frame" up into the application layer (e.g. override `Http2ConnectionHandler#onStreamError`) before considering baking this into Netty.

In reference to @ejona86's comment https://github.com/netty/netty/issues/8025#issuecomment-397689321 IIUC his suggestion is to ignore the receipt of RST_STREAM on a closed stream, but we already do that [1][2].

[1] https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L341-L342
[2] https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L349

Reply:
> If the stream may have existed ... tracking/estimating RTT time is error prone and can be expensive for what is in large part a boundary condition. The RFC permits sending multiple RST_STREAM frames, doing so will have no negative implications from a protocol perspective*. A simpler approach would be to continue to send the RST_STREAM frames which provides positive acknowledgement for good peers that this frame was not accepted, and have some heuristic to "fail hard" (e.g. connection error) to catch the case of "misbehaving peers".

How would we trigger a RST_STREAM without throwing a stream error?

Reply:
@Scottmitch can you take a look at my latest comment?

Reply:
sorry was afk for a few days

> How would we trigger a RST_STREAM without throwing a stream error?

throwing a `StreamException` in the interface based API (e.g. `Http2ConnectionHandler`, `Http2ConnectionDecoder`, `Http2ConnectionEncoder`, etc..) will generally result in a RST_STREAM being sent. Independent of this you can also always call [writeRstStream](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameWriter.java#L121) on the `Http2ConnectionDecoder`.

The code I provided in the previous comment [1] https://github.com/netty/netty/pull/8028#discussion_r197092876 should be sufficient to cover the "If the stream may have existed" and "if the stream was never in existence" case. However this does't cover the heuristic to close the connection if receiving a frame for a stream in a closed state should be a protocol error instead of a stream error, and I think we can handle that as a followup PR.

[1]
```java
if (connection.streamMayHaveExisted(streamId)) {
  throw streamError(streamId, STREAM_CLOSED, "Received %s frame for an unknown stream %d", frameName, streamId);
} else {
  throw connectionError(PROTOCOL_ERROR, "Stream %d does not exist", streamId);
}
```